usr/lib/common/pkey_utils.c

```diff
@@ -0,0 +1,531 @@
+ */
+static int s390_pcc(unsigned long func, void *param)
+{
+    register unsigned long r0 __asm__("0") = (unsigned long)func;
+    register unsigned long r1 __asm__("1") = (unsigned long)param;
+
+    __asm__ volatile (
+        "0: .long   %[opc] << 16\n"
+        "   brc 1,0b\n"
+        :
+        : [fc] "d" (r0), [param] "a" (r1), [opc] "i" (0xb92c)
+        : "cc", "memory");
+
+    return 0;
+}
+
+/**
+ * Update the specified attribute of the given key object. The object gets
+ * locked for write and is saved, if it's a token object.
+ *
+ * Note: When calling this function, the XProcLock MUST NOT be held,
+ *       because it tries to obtain a write lock on the key object.
+ */
+CK_RV pkey_update_and_save(STDLL_TokData_t *tokdata, OBJECT *key_obj,
+                           CK_ATTRIBUTE *pkey_attr)
+{
+    CK_RV ret;
+
+    /* Unlock obj to remove READ_LOCK */
+    ret = object_unlock(key_obj);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_unlock failed with rc=0x%lx\n", ret);
+        return ret;
+    }
+
+    /* Obtain write lock on key object */
+    ret = object_lock(key_obj, WRITE_LOCK);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_lock for WRITE failed with rc=0x%lx\n", ret);
+        /* Try to restore read lock without rc checking */
+        object_lock(key_obj, READ_LOCK);
+        return ret;
+    }
+
+    /* Update attribute */
+    ret = template_update_attribute(key_obj->template, pkey_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed with rc=0x%lx\n", ret);
+        goto done;
+    }
+
+    /* Save to repository if it's a token object */
+    if (object_is_token_object(key_obj)) {
+        ret = object_mgr_save_token_object(tokdata, key_obj);
+        if (ret != CKR_OK) {
+            TRACE_ERROR("Could not save token obj to repository, rc=0x%lx.\n", ret);
+        }
+    }
+
+done:
+
+    /* Restore object's READ_LOCK */
+    ret = object_unlock(key_obj);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_unlock failed with rc=0x%lx\n", ret);
+    } else {
+        ret = object_lock(key_obj, READ_LOCK);
+        if (ret != CKR_OK)
+            TRACE_ERROR("object_lock for READ failed with rc=0x%lx\n", ret);
+    }
+
+    return ret;
+}
+
+/**
+ * Returns true if the protected key operation implied by the given mechanism
+ * is supported by CPACF, false otherwise.
+ */
+CK_BBOOL pkey_op_supported_by_cpacf(CK_MECHANISM *mech)
+{
+    if (!mech)
+        return CK_FALSE;
+
+    switch (mech->mechanism) {
+    case CKM_AES_ECB:
+    case CKM_AES_CBC:
+    case CKM_AES_CBC_PAD:
+    case CKM_AES_CMAC_GENERAL:
+    case CKM_AES_CMAC:
+        return CK_TRUE;
+    default:
+        return CK_FALSE;
+    }
+}
+
+/**
+ * Return the KM/KMC/KMAC function code for the given key length and
+ * encrypt/decrypt op.
+ */
+unsigned long get_function_code(CK_ULONG clear_keylen, CK_BYTE encrypt)
+{
+    unsigned long fc;
+
+    switch (clear_keylen) {
```

Comment:
how about `fc = ((clear_keylen - 16) >> 3) + 0x1a`?

Reply:
well, does this really making it easier to read?  :-)

Reply:
Hm yes .... I would prefer it this way ...

Reply:
Focus on readability and leave optimizations to the compiler ... I would only consider doing that kind of stuff if its in an absolutely performance critical code path.